### ISSUE TITLE
TRGP: Trust Region Gradient Projection for Continual Learning

### ISSUE THREAD

**Issue by chaerinkong** [reyllama/MIPAL-CL#5]
## TL; DR

- Combine orthogonal gradient projection with subspace scaling to better handle task correlations.

## Motivation

- Orthogonal gradient projection fails when two tasks are highly correlated, in which case simple linear operations within the subspace can provide meaningful knowledge.

### Problem Statement

![image](https://user-images.githubusercontent.com/46308407/158134470-f85cc8f2-80d6-457e-ab7f-2b829c31274c.png)

### Layer-wise Trust Region

- Set of previous tasks with high correlation with the current one. (Graphical illustration below)

![image](https://user-images.githubusercontent.com/46308407/158134543-164e119d-4d2a-418e-afcb-6d938e6c95a1.png)

### Scaled Weight Projection

- Keep additional scaling matrix for each layer, which is later used for linear scaling within the trust region.

![image](https://user-images.githubusercontent.com/46308407/158134628-33715d14-b52a-45bf-a5fe-c9c6d021c309.png)

### Subspace construction

- Forward pass to obtain layer-wise representations, and apply SVD.

## Evaluation

![image](https://user-images.githubusercontent.com/46308407/158134875-639f9e9b-c892-4860-b9da-47f11f0cdd33.png)
![image](https://user-images.githubusercontent.com/46308407/158134903-d919e366-efd1-4064-a93f-5904c2baec00.png)
![image](https://user-images.githubusercontent.com/46308407/158134925-e2f687d5-406a-4541-8643-91005fa6034e.png)
![image](https://user-images.githubusercontent.com/46308407/158134937-8f4908db-2f37-45c5-b3aa-cb1a350a2131.png)
![image](https://user-images.githubusercontent.com/46308407/158134960-5b46e4ba-e993-421c-b3b3-1a78764e62cd.png)
![image](https://user-images.githubusercontent.com/46308407/158134973-2a37ac3c-2bd2-4a74-b264-72c959e8d695.png)

## Comments

- Clever idea to leverage highly correlated tasks through simple scaling in the task subspace.
